third_party/terraform/resources/resource_dataflow_job.go

```diff
@@ -121,44 +126,78 @@ func resourceDataflowJob() *schema.Resource {
 			"service_account_email": {
 				Type:     schema.TypeString,
 				Optional: true,
-				ForceNew: true,
 			},
 
 			"network": {
 				Type:             schema.TypeString,
 				Optional:         true,
-				ForceNew:         true,
 				DiffSuppressFunc: compareSelfLinkOrResourceName,
 			},
 
 			"subnetwork": {
 				Type:             schema.TypeString,
 				Optional:         true,
-				ForceNew:         true,
 				DiffSuppressFunc: compareSelfLinkOrResourceName,
 			},
 
 			"machine_type": {
 				Type:     schema.TypeString,
 				Optional: true,
-				ForceNew: true,
 			},
 
 			"ip_configuration": {
 				Type:         schema.TypeString,
 				Optional:     true,
-				ForceNew:     true,
 				ValidateFunc: validation.StringInSlice([]string{"WORKER_IP_PUBLIC", "WORKER_IP_PRIVATE", ""}, false),
 			},
 
 			"job_id": {
 				Type:     schema.TypeString,
 				Computed: true,
 			},
+
+			// Detect external changes made to the template storage object
+			"template_md5hash": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
 		},
 	}
 }
 
+func resourceDataflowJobTypeCustomizeDiff(d *schema.ResourceDiff, meta interface{}) error {
+	config := meta.(*Config)
+
+	if d.HasChange("template_gcs_path") {
+		d.SetNewComputed("template_md5hash")
+	} else {
+		// Detect change in template hash
+		oldmd5Hash := d.Get("template_md5hash")
+		newMd5Hash, err := resourceDataflowJobReadTemplateStorageObjectHash(d.Get("template_gcs_path").(string), config)
+		if err != nil {
+			return err
+		}
+		if oldmd5Hash != "" && oldmd5Hash != newMd5Hash {
+			d.SetNew("template_md5hash", newMd5Hash)
+		}
+	}
+
+	// All changes are ForceNew for batch jobs
+	if d.Get("type") == "JOB_TYPE_BATCH" {
+		resourceSchema := resourceDataflowJob().Schema
```

Comment:
Couldn't find another way to get all of the resource keys. `ResourceDiff` doesn't provide access to the schema. If there is an alternative way to be able to iterate and set any change to ForceNew, I'd be happy to hear it

Reply:
The only other way I can think of would be to create a top-level var for the schema that gets referenced from both places (kind of like what we do for node pool),  but the overall approach would be the same.